js/src/main/scala/com/raquo/domtypes/jsdom/defs/package.scala

```diff
@@ -1,20 +1,26 @@
 import org.scalajs.dom
 
 package object defs {
 
   object eventProps {
     type ClipboardEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.ClipboardEventProps[EP, dom.Event, dom.ClipboardEvent]
     type ErrorEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.ErrorEventProps[EP, dom.Event, dom.ErrorEvent]
-    type FormEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.FormEventProps[EP, dom.Event, dom.FocusEvent, dom.Event, ElementTargetEvent[dom.html.Element], ElementTargetEvent[dom.html.Form], ElementTargetEvent[dom.html.Input]]
+    type FormEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.FormEventProps[EP, dom.Event, dom.FocusEvent, dom.Event, TypedTargetEvent[dom.html.Element], TypedTargetEvent[dom.html.Form], TypedTargetEvent[dom.html.Input]]
     type KeyboardEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.KeyboardEventProps[EP, dom.Event, dom.KeyboardEvent]
     type MediaEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.MediaEventProps[EP, dom.Event]
-    type MiscellaneousEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.MiscellaneousEventProps[EP, dom.Event]
-    type MouseEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.MouseEventProps[EP, dom.Event, dom.MouseEvent, dom.DragEvent]
-    type WindowEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.WindowEventProps[EP, dom.Event, dom.UIEvent, dom.BeforeUnloadEvent, dom.HashChangeEvent, PageTransitionEvent, dom.PopStateEvent, dom.StorageEvent]
+    type MiscellaneousEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.MiscellaneousEventProps[EP, dom.Event, dom.UIEvent]
+    type MouseEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.MouseEventProps[EP, dom.Event, dom.MouseEvent, dom.DragEvent, TypedTargetEvent[dom.html.Element]]
+    type WindowEventProps[EP[_ <: dom.Event]] = generic.defs.eventProps.WindowEventProps[EP, dom.Event, dom.UIEvent, dom.BeforeUnloadEvent, dom.HashChangeEvent, dom.MessageEvent, PageTransitionEvent, dom.PopStateEvent, dom.StorageEvent]
+
+    trait EventPropsInGlobal[EP[_ <: dom.Event]] extends ErrorEventProps[EP] with FormEventProps[EP] with KeyboardEventProps[EP] with MediaEventProps[EP] with MiscellaneousEventProps[EP] with MouseEventProps[EP] { this: generic.builders.EventPropBuilder[EP, dom.Event] => }
+    trait EventPropsInWindow[EP[_ <: dom.Event]] extends EventPropsInGlobal[EP] with WindowEventProps[EP] { this: generic.builders.EventPropBuilder[EP, dom.Event] => }
+    trait EventPropsInDocument[EP[_ <: dom.Event]] extends EventPropsInGlobal[EP] with ClipboardEventProps[EP] { this: generic.builders.EventPropBuilder[EP, dom.Event] => }
+    trait EventPropsInElement[EP[_ <: dom.Event]] extends EventPropsInGlobal[EP] { this: generic.builders.EventPropBuilder[EP, dom.Event] => }
+    trait EventPropsInHTMLElement[EP[_ <: dom.Event]] extends EventPropsInGlobal[EP] with ClipboardEventProps[EP] { this: generic.builders.EventPropBuilder[EP, dom.Event] => }
```

Comment:
Hmm... I'm not sure if these new traits should reside in SDT. I don't really see a good way to use more than one of those traits at a time in a typical consuming library. I mean, the names would conflict, and you'd need to instantiate way more eventprop objects than needed... right? Maybe I'm missing something?

I'm guessing #13 is related here

Reply:
Actually, I would like sdt to bring those type as they tell me which event properties are applicable to which underlying `EventTarget` type.

My plan is to have the following three usages:
- `trait DomEventProps extends EventPropsInHTMLElement[EmitterBuilder] with EventPropBuilder[EmitterBuilder, dom.Event]` for normal usage within tags as it is now, like: `input(onChange --> sink)`.
- `object WindowEvents extends EventPropsInWindow[Observable] with EventPropBuilder[Observable, dom.Event]` for registering window events via an observable: `WindowEvents.onStorage.map(ev => ...)`.
- `object DocumentEvents extends EventPropsInDocument[Observable] with EventPropBuilder[Observable, dom.Event]` for registering document events via an observable: `DocumentEvents.onLoad.map(ev => ...)`.

The latter two can be build by creating an observable from the callback we have when registering an event with `(Window|Document).attachEventListener(eventKey, callback)`

Reply:
Oh that's a nice pattern! I myself was using `node.events(onClick)` but I don't want each node to extend a trait with a hundred vals, even if they're lazy. But for document / window this looks reasonable.

Let's keep all those traits then.

---

I'm assuming you named the new traits `...InX` as opposed to `X...` because `EventPropsInWindow` would otherwise conflict with `WindowEventProps`? Should we maybe rename the original `WindowEventProps` to `WindowOnlyEventProps` or something like that? Would look nicer and would also sort of better match Javascript's stuff:

https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers
https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers

Reply:
Exactly! Sure, we can rename them; I wasn't too happy with this `...In...` :)

Reply:
> node.events(onClick)

With regards to https://github.com/OutWatch/outwatch/issues/93, that approach would work much better with the `currentTarget`-event.